_posts/2024-08-19-given-priority-change-3.7.md

```diff
@@ -0,0 +1,232 @@
+
+```scala
+trait Functor[F[_]]:
+  extension [A, B](x: F[A]) def map(f: A => B): F[B]
+trait Monad[F[_]] extends Functor[F] { ... }
+trait Traverse[F[_]] extends Functor[F] { ... }
+```
+Now, suppose we have corresponding instances of these typeclasses for `List`:
+```scala
+given Functor[List]  = ...
+given Monad[List]    = ...
+given Traverse[List] = ...
+```
+Let’s use these in the following context:
+```scala
+def fmap[F[_] : Functor, A, B](c: F[A])(f: A => B): F[B] = c.map(f)
+
+fmap(List(1,2,3))(_.toString)
+// ^ rejected by Scala < 3.7, now accepted by Scala 3.7
+```
+
+Before Scala 3.7, the compiler would reject the `fmap` call due to
+ambiguity. Since it prioritized the `given` instance with the most
+specific subtype of the context bound `Functor`, both `Monad[List]` and
+`Traverse[List]` were valid candidates for `Functor[List]`, but neither
+was more specific than the other. However, all that’s required is the
+functionality of `Functor[List]`, the _most general_ instance, which Scala
+3.7 correctly picks.
+
+This change aligns the behavior of the compiler with the practical
+needs of developers, making the handling of common triangle inheritance
+patterns more predictable.
+
+## Migrating to the New Prioritization
+
+### Community Impact
+
+We have conducted experiments on the [open community
+build](https://github.com/VirtusLab/community-build3) that showed that
+the proposed scheme will result in a more intuitive and predictable
+given resolution. The negative impact on the existing projects is very
+small. We have tested 1500 open-source libraries, and new rules are
+causing problems for less than a dozen of them.
+
+### Roadmap
+
+The new `given` resolution scheme, which will be the default in Scala
+3.7, can already be explored in Scala 3.5. This early access allows
+the community ample time to test and adapt to the upcoming changes.
+
+**Scala 3.5**
+
+Starting with Scala 3.5, you can compile with `-source 3.6` to receive
+warnings if the new `given` resolution scheme would affect your
+code. This is how the warning might look:
+
+```scala
+-- Warning: client.scala:11:30 ------------------------------------------
+11 |@main def run = printComponent
+   |                              ^
+   |           Given search preference for LibComponent between alternatives
+   |             (userComponent : UserComponent)
+   |           and
+   |             (libComponent : LibComponent)
+   |           has changed.
+   |           Previous choice          : the first alternative
+   |           New choice from Scala 3.7: the second alternative
+```
+
+Additionally, you can compile with `-source 3.7` or `-source future`
+to fully enable the new prioritization and start experiencing its
+effects.
+
+**Scala 3.6**
+
+In Scala 3.6, these warnings will be on by default.
+
+**Scala 3.7**
+
+Scala 3.7 will finalize the transition, making the new given
+prioritization the standard behavior.
+
+#### Suppressing Warnings
```

Comment:
We can consider adding a paragraph here that suppressing warnings is only a temporary solution. Eg. 

```
Consider that suppressing warnings should become only a temporary workaround. 
They can be used to resolve rare false-positive warnings produced by the compiler. 
However, they're only delaying the time when changes in either your codebase or libraries used by your project are required. Suppressing warnings can lead to unexpected problems when upgrading to the future versions of the Scala compiler.
```

Reply:
Done!